### DISCUSSION TITLE
Is there a trick on getServerSideProps to check if user types the URL directly

### DISCUSSION THREAD

**Discussion by cha2hyun** [vercel/next.js#45077] [Help] [answered]
### About
1. When user clicks post on the list. API give every details of post as an array.
2. So the user doesnt need to fetch again when entering the post page.
3. However, if user go directly to the post page without going through the list page, an error occurs because there are no props
4. So i use `useEffect` to check the condition to see if the props doesnt exist's and get the data from API 

### Problem
**'SEO'** didn't work because the page renders after providing SEO data. so meta tag says nothing instead of post information (title, author, content)

### Question
Is there a trick on getServerSideProps to check if user types the URL directly 

### Code [id].tsx
```ts
...
  React.useEffect(() => {
    if (!router.query.post) {
      getPostDetail(router.asPath.split('/')[3]).then(function (response) {
        setPost(response.results[0]);
      });
    } else {
      setPost(JSON.parse(router.query.post as string));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
...
```

**Comment by icyJoseph** [comment, marked as answer]
I am not sure I understand where's `getServerSideProps` involved on this?

Who uses `getServerSIdeProps`? 

You have a page A: shows list of posts.

You have a page B: shows post details

Users can navigate from A to B, and see post details. 

Going directly to B causes problems, because there's no data for the post.

So you want to use `getServerSideProps` on page B, so that there's data when the user arrives there, but avoid fetching it from page A, because you already have the data? But how did page B get data from page A in the first page?

**Reply by cha2hyun** [reply]
Thanks for reply. 

Lets say
Page A : List of posts
Page B : Post Details

## Case 1 : Navigate Page A to Page B 
Send post's details as query to Page B. User doesn't need to fetch any data.

```ts
Page A 
{data?.pages.map((page) =>
  page.results.map((post: Results) => (
    <li key={post.id} data-aos='fade-up' data-aos-duration='200'>
      <Link
        href={{
          pathname: `community/post/${post.id}`,
          query: { post: JSON.stringify(post) },
        }}
        as={`community/post/${post.id}`}
        passHref
      >
        <a
          href={`community/post/${post.id}`}
          onClick={() => setScrollY(window.scrollY)}
        >
...
```

## Case 2 : Navigate directly to Page B

It occurs error because there is no Post's details.
So using with useEffect to check `router.query` has data or not. if not it fetch data from API
```ts
Page B
...
  React.useEffect(() => {
    if (!router.query.post) {
      getPostDetail(router.asPath.split('/')[3]).then(function (response) {
        setPost(response.results[0]);
      });
    } else {
      setPost(JSON.parse(router.query.post as string));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
...
```

## Result
The code works well. Case1 Case2 works well

## Problem 
But the problem is `SEO metatag` (I have to fetch data before page render) (Case2)
So i'd like to change Page B's `useEffect` to `getServerSideProps` for `SEO metatag` with check if there is query data or not

🙏

**Reply by icyJoseph** [reply]
Right, the solution is to indeed use `getServerSideProps` - or - pre-render every possible post at build time with `getStaticPaths` + `getStaticProps`, which also opens up ISR.

Anyway, passing data through the queries is not ideal. People could inject their own data that way, and those links could be shared, and they'll render stuff that was injected through the query. Be-careful with that.

Using `getServerSideProps` means that from the client, when the user navigates with next/link, a JSON payload is used as page props, but when the user navigates directly to your page, the hole page is generated server side, so for example, meta tags are present. 

It might feel a bit bad because you already have the post data in page A, and you are forced to refetch it in the server, and send it again, but then you should probably reduce how much data page A gets to see from each post, to the minimum necessary.

You might also loose speed, passing the data with queries is most likely faster than the round trip with `getServerSideProps`...

## Getting Hacky...

This is a bit of a hack that works... but might take some time to get around:

- Create `pages/blog/[[...slug]].tsx` with double`[[...]]`

```tsx
import { GetServerSidePropsContext } from "next";
import Link from "next/link";
import { useRouter } from "next/router";
import { useCallback } from "react";

type BlogEntry = {
  id: string;
  title: string;
  content: string;
};

type BlogProps = {
  posts: BlogEntry[];
};

const Blog = ({ posts }: BlogProps) => {
  return (
    <>
      <h1>Blog</h1>

      <ul>
        {posts.map((post) => (
          <li key={post.id}>
            <Link
              href={`/blog/${post.id}`}
              shallow
              style={{ color: "palevioletred" }}
            >
              {post.title}
            </Link>
            - {post.content}
          </li>
        ))}
      </ul>
    </>
  );
};

const Post = ({ id, title, content }: BlogEntry) => {
  return (
    <>
      <h1>{title}</h1>

      <p>{content}</p>

      <div>
        <Link href="/blog" shallow style={{ color: "palevioletred" }}>
          Back
        </Link>
      </div>
    </>
  );
};

const BlogPage = ({ posts }: BlogProps) => {
  const router = useRouter();

  const refreshData = () => {
    router.replace(router.asPath);
  };

  const { slug } = router.query;

  const UI = useCallback(() => {
    if (Array.isArray(slug) && slug.length === 1) {
      const post = posts.find((p) => p.id === slug[0]);

      // sanity check
      // - Blog renders a list of existing links
      // - Navigating to post directly, passes through GSSP
      //   and returns 404 for non existing posts
      if (!post) return <h1>404</h1>;

      return <Post {...post} />;
    }

    return <Blog posts={posts} />;
  }, [posts, slug]);

  return (
    <>
      <UI />
      <button onClick={refreshData}>Refresh data</button>
    </>
  );
};

// ignore this part, it is just to make random strings
const randomStr = () =>
  Date.now().toString(36) + Math.random().toString(36).slice(2);
const getData = () => [
  { id: "0", title: "foo", content: randomStr() },
  { id: "1", title: "joe", content: randomStr() },
  { id: "2", title: "hello", content: randomStr() },
]; // somehow generating random data

// GSSP
export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  const slug = ctx.params?.slug;

  const data = getData();
  // arriving at /blog
  if (!slug) return { props: { posts: data } };
  if (!Array.isArray(slug)) return { notFound: true };

  // only allow slug = [<post-id>]
  if (slug.length !== 1) return { notFound: true };

  const post = data.find((p) => p.id === slug[0]);

  if (!post) return { notFound: true };

  return { props: { posts: data } };
};

export default BlogPage;
```

See it in action:


https://user-images.githubusercontent.com/21013447/213705529-f1dcdeff-35ba-4085-9167-1d14b61228b2.mov

**Reply by cha2hyun** [reply]
Best answer ever
Thanks for the tips 🙏

**Comment by GabenGar** [comment]
There is no way to "check" if the URL was typed directly sans client-only navigation. But relying on client navigation means it will break in hybrid contexts.
You should write your page components as isolated entities which derive their state from the URL alone, with other sources (cookies, `localStorage` etc.) being of secondary value.

**Reply by cha2hyun** [reply]
Thanks !
